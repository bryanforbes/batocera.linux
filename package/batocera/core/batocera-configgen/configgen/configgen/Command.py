from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence
    from pathlib import Path


class Command:
    def __init__(self, array: Sequence[str | Path], env: Mapping[str, str | Path] = {}):
        self.array = list(array)
        self.env = dict(env)

    def __str__(self):
        strings = [f"{varName}={varValue}" for varName, varValue in self.env.items()]

        strings.extend(str(value) for value in self.array)

        return " ".join(strings)
