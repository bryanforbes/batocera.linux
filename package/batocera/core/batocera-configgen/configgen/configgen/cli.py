from __future__ import annotations

import logging
import time
from collections.abc import Callable
from functools import partial
from pathlib import Path
from typing import Any, TypedDict, TypeVar, Unpack, cast

import click

from . import profiler
from .emulators import run_emulator
from .utils.logger import setup_logging

_FC = TypeVar('_FC', bound=Callable[..., Any] | click.Command)

_logger = logging.getLogger(__name__)


def _player_callback(player: int, ctx: click.Context, option: click.Option, value: int | None) -> int | None:
    current_value: list[PlayerOptions] = ctx.params.setdefault('players', [])

    option_names = (
        f'p{player}hat_count',
        f'p{player}button_count',
        f'p{player}device_path',
        f'p{player}real_name',
        f'p{player}guid',
        f'p{player}index',
    )

    items = [
        (f'p{player}axis_count', value),
        *[(option_name, ctx.params.pop(option_name)) for option_name in option_names],
    ]

    if all(item[1] is not None for item in items):
        p_length = len(f'p{player}')
        current_value.append(
            cast(
                PlayerOptions,
                {
                    'player_number': player,
                    **{option_name[p_length:]: option_value for option_name, option_value in items},
                },
            )
        )
        return value

    if all(item[1] is None for item in items):
        return value

    raise click.UsageError(f'All options that start with p{player} must be specified if any are specified', ctx=ctx)


def _players(max_players: int) -> Callable[[_FC], _FC]:
    def wrap(callable: _FC) -> _FC:
        for n in range(max_players, 0, -1):
            callable = click.option(f'-p{n}nbaxes', f'p{n}axis_count', type=int, callback=partial(_player_callback, n))(
                callable
            )
            callable = click.option(f'-p{n}nbhats', f'p{n}hat_count', type=int)(callable)
            callable = click.option(f'-p{n}nbbuttons', f'p{n}button_count', type=int)(callable)
            callable = click.option(f'-p{n}devicepath', f'p{n}device_path', type=str)(callable)
            callable = click.option(f'-p{n}name', f'p{n}real_name', type=str)(callable)
            callable = click.option(f'-p{n}guid', type=str)(callable)
            callable = click.option(f'-p{n}index', type=int)(callable)

        return callable

    return wrap


class PlayerOptions(TypedDict):
    player_number: int
    index: int
    guid: str
    real_name: str
    device_path: str
    button_count: int
    hat_count: int
    axis_count: int


class Options(TypedDict):
    system: str
    rom: Path
    emulator: str | None
    core: str | None
    netplaymode: str | None
    netplaypass: str | None
    netplayip: str | None
    netplayport: str | None
    netplaysession: str | None
    state_slot: str | None
    state_filename: Path | None
    autosave: str | None
    systemname: str | None
    gameinfoxml: Path
    lightgun: bool
    wheel: bool
    trackball: bool
    spinner: bool
    players: list[PlayerOptions]


@click.command()
@click.option('-system', required=True)
@click.option('-rom', required=True, type=click.Path(path_type=Path))
@click.option('-emulator')
@click.option('-core')
@click.option('-netplaymode')
@click.option('-netplaypass')
@click.option('-netplayip')
@click.option('-netplayport')
@click.option('-netplaysession')
@click.option('-state_slot')
@click.option('-state_filename', type=click.Path(path_type=Path))
@click.option('-autosave')
@click.option('-systemname')
@click.option('-gameinfoxml', default='/dev/null', type=click.Path(path_type=Path))
@click.option('-lightgun', is_flag=True)
@click.option('-wheel', is_flag=True)
@click.option('-trackball', is_flag=True)
@click.option('-spinner', is_flag=True)
@_players(8)
def cli(**kwargs: Unpack[Options]) -> None:
    with setup_logging():
        exit_code = -1

        try:
            exit_code = run_emulator(kwargs)
        except Exception:
            _logger.exception('configgen exception: ')

        profiler.disable(dump=True)

        time.sleep(1)  # this seems to be required so that the gpu memory is restituated and available for es
        _logger.debug('Exiting configgen with status %s', exit_code)

        exit(exit_code)
