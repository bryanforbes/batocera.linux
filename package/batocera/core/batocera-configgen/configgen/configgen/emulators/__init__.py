from __future__ import annotations

from contextlib import nullcontext
from importlib import import_module
from typing import TYPE_CHECKING, Final

from ..batoceraPaths import BATOCERA_CONF
from ..config import RenderConfig, SystemConfig
from ..controller import Controller
from ..controllersConfig import getGamesMetaData
from ..gun import Gun, precalibrate_guns
from ..settings.unixSettings import UnixSettings
from ..utils.squashfs import squashfs_rom
from ..utils.wheels import configure_wheels
from ._base import Emulator as Emulator  # noqa: TCH001

if TYPE_CHECKING:
    from ..cli import Options

_GENERATOR_MAP: Final[dict[str, str]] = {}


def get_emulator_cls(emulator: str, /) -> type[Emulator]:
    module_path = _GENERATOR_MAP.get(emulator, emulator.replace('-', '_'))

    try:
        module = import_module(f'..{module_path}', package=__name__)
        emulator_cls: type[Emulator] = module.Emulator
    except ImportError as e:
        if e.name is not None and e.name.startswith(__name__.split('.')[0]):
            raise Exception(f'no emulator found for {emulator!r}') from e

        raise
    except AttributeError as e:
        raise Exception(f'no emulator found for {emulator!r}') from e

    return emulator_cls


def run_emulator(options: Options, /) -> int:
    exit_code = -1
    original_rom = options['rom']

    if original_rom.suffix == '.squashfs':
        rom_context_manager = squashfs_rom(original_rom)
    else:
        rom_context_manager = nullcontext(original_rom)

    with rom_context_manager as rom:
        system = options['system']

        settings = UnixSettings(BATOCERA_CONF)
        system_config = SystemConfig.load(options, settings)
        render_config = RenderConfig.load(options, settings, system_config.get('shaderset'))

        metadata = getGamesMetaData(system, original_rom)

        controllers = Controller.load_for_player_options(options['players'])

        if system_config.use_guns:
            guns = Gun.get_all()
            precalibrate_guns(system, system_config.emulator, system_config.core, original_rom)
        else:
            guns = {}

        with configure_wheels(controllers, system, system_config, metadata) as (controllers, wheels):
            emulator_cls = get_emulator_cls(system_config.emulator)
            emulator = emulator_cls(system, system_config, render_config, metadata, controllers, guns, wheels)
            exit_code = emulator.run(rom, original_rom, options['gameinfoxml'])

    return exit_code
