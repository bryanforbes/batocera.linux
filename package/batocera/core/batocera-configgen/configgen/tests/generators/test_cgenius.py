from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING, Any

import pytest

from configgen.batoceraPaths import CONFIGS, ROMS
from configgen.config import SystemConfig
from configgen.generators.cgenius.cgeniusGenerator import CGeniusGenerator
from tests.generators.base import GeneratorBaseTest

if TYPE_CHECKING:
    from pyfakefs.fake_filesystem import FakeFilesystem
    from syrupy.assertion import SnapshotAssertion

    from configgen.controller import Controller, Controllers
    from configgen.Emulator import Emulator


class TestCGeniusGenerator(GeneratorBaseTest):
    @pytest.fixture
    def generator_cls(self) -> type[CGeniusGenerator]:
        return CGeniusGenerator

    @pytest.fixture
    def system_name(self) -> str:
        return 'cgenius'

    @pytest.fixture
    def emulator(self) -> str:
        return 'cgenius'

    @pytest.fixture(autouse=True)
    def fs(self, fs: FakeFilesystem) -> FakeFilesystem:
        fs.create_dir(ROMS / 'cgenius')

        return fs

    def test_get_mouse_mode(self, generator: CGeniusGenerator) -> None:  # pyright: ignore
        assert generator.getMouseMode(SystemConfig({}), Path())

    @pytest.mark.parametrize(
        ('mock_system_config', 'result'),
        [
            ({}, 4 / 3),
            ({'cgenius_aspect': '5:4'}, 4 / 3),
            ({'cgenius_aspect': '16:9'}, 16 / 9),
            ({'cgenius_aspect': '16:10'}, 16 / 9),
        ],
        ids=str,
    )
    def test_get_in_game_ratio(  # pyright: ignore
        self, generator: CGeniusGenerator, mock_system_config: dict[str, Any], result: bool
    ) -> None:
        assert generator.getInGameRatio(SystemConfig(mock_system_config), {'width': 0, 'height': 0}, Path()) == result

    def test_generate(
        self,
        generator: CGeniusGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'cgenius' / 'rom.cgenius',
                one_player_controllers,
                {},
                [],
                {},
                {},  # pyright: ignore
            )
            == snapshot
        )
        assert (CONFIGS / 'cgenius' / 'cgenius.cfg').read_text() == snapshot(name='config')
        assert (ROMS / 'cgenius' / 'cgenius.cfg').read_text() == (CONFIGS / 'cgenius' / 'cgenius.cfg').read_text()

    def test_generate_existing(
        self,
        generator: CGeniusGenerator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_file(
            CONFIGS / 'cgenius' / 'cgenius.cfg',
            contents="""[FileHandling]
EnableLogfile = false
SearchPath1 = /userdata/roms/cgenius
SearchPath2 = /userdata/roms/cgenius/games
[Video]
aspect = 4:3
fullscreen = false
integerScaling = false
filter = none
OGLfilter = nearest
gameHeight = 200
gameWidth = 320
ShowCursor = false
[input0]
Fire = Joy0-B3
Jump = Joy0-B2
Down = Joy0-H1
Left = Joy0-H7
Run = Joy0-B4
Camlead = Joy0-B8
Right = Joy0-H3
Up = Joy0-H3
Status = Joy0-B1
Pogo = Joy0-B9
""",
        )

        assert (
            generator.generate(
                mock_system,
                ROMS / 'cgenius' / 'rom.cgenius',
                one_player_controllers,
                {},
                [],
                {},
                {},  # pyright: ignore
            )
            == snapshot
        )
        assert (CONFIGS / 'cgenius' / 'cgenius.cfg').read_text() == snapshot(name='config')
        assert (ROMS / 'cgenius' / 'cgenius.cfg').read_text() == (CONFIGS / 'cgenius' / 'cgenius.cfg').read_text()

    @pytest.mark.parametrize(
        'mock_system_config',
        [
            {'cgenius_aspect': '16:9'},
            {'cgenius_filter': 1},
            {'cgenius_quality': 'best'},
            {'cgenius_render': '200'},
            {'cgenius_render': '240'},
            {'cgenius_render': '360'},
            {'cgenius_render': '480'},
            {'cgenius_cursor': 'true'},
        ],
        ids=str,
    )
    def test_generate_config(
        self,
        generator: CGeniusGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'cgenius' / 'rom.cgenius',
                one_player_controllers,
                {},
                [],
                {},
                {},  # pyright: ignore
            )
            == snapshot
        )
        assert (CONFIGS / 'cgenius' / 'cgenius.cfg').read_text() == snapshot(name='config')
        assert (ROMS / 'cgenius' / 'cgenius.cfg').read_text() == (CONFIGS / 'cgenius' / 'cgenius.cfg').read_text()

    def test_generate_controllers(
        self,
        generator: CGeniusGenerator,
        mock_system: Emulator,
        generic_xbox_pad: Controller,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'cgenius' / 'rom.cgenius',
                [
                    generic_xbox_pad.replace(
                        player_number=n, index=n - 1, real_name=f'real name {n}', device_path=f'/dev/input/event{n}'
                    )
                    for n in range(1, 6)
                ],
                {},
                [],
                {},
                {},  # pyright: ignore
            )
            == snapshot
        )
        assert (CONFIGS / 'cgenius' / 'cgenius.cfg').read_text() == snapshot(name='config')
        assert (ROMS / 'cgenius' / 'cgenius.cfg').read_text() == (CONFIGS / 'cgenius' / 'cgenius.cfg').read_text()
