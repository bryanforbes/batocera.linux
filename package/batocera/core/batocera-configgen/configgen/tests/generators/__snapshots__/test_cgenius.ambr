# serializer version: 1
# name: TestCGeniusGenerator.test_generate
  tuple(
    list([
      'CGeniusExe',
      'dir="."',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestCGeniusGenerator.test_generate[config]
  '''
  [FileHandling]
  EnableLogfile = false
  SearchPath1 = /userdata/roms/cgenius
  SearchPath2 = /userdata/roms/cgenius/games
  [Video]
  aspect = 4:3
  fullscreen = false
  integerScaling = false
  filter = none
  OGLfilter = nearest
  gameHeight = 200
  gameWidth = 320
  ShowCursor = false
  [input0]
  Fire = Joy0-B0
  Jump = Joy0-B1
  Down = Joy0-H4
  Left = Joy0-H8
  Run = Joy0-B5
  Camlead = Joy0-B4
  Right = Joy0-H2
  Up = Joy0-H1
  Status = Joy0-B2
  Pogo = Joy0-B3
  
  '''
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_aspect': '16:9'}]
  tuple(
    list([
      'CGeniusExe',
      'dir="."',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_aspect': '16:9'}][config]
  '''
  [FileHandling]
  EnableLogfile = false
  SearchPath1 = /userdata/roms/cgenius
  SearchPath2 = /userdata/roms/cgenius/games
  [Video]
  aspect = 16:9
  fullscreen = false
  integerScaling = false
  filter = none
  OGLfilter = nearest
  gameHeight = 200
  gameWidth = 320
  ShowCursor = false
  [input0]
  Fire = Joy0-B0
  Jump = Joy0-B1
  Down = Joy0-H4
  Left = Joy0-H8
  Run = Joy0-B5
  Camlead = Joy0-B4
  Right = Joy0-H2
  Up = Joy0-H1
  Status = Joy0-B2
  Pogo = Joy0-B3
  
  '''
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_cursor': 'true'}]
  tuple(
    list([
      'CGeniusExe',
      'dir="."',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_cursor': 'true'}][config]
  '''
  [FileHandling]
  EnableLogfile = false
  SearchPath1 = /userdata/roms/cgenius
  SearchPath2 = /userdata/roms/cgenius/games
  [Video]
  aspect = 4:3
  fullscreen = false
  integerScaling = false
  filter = none
  OGLfilter = nearest
  gameHeight = 200
  gameWidth = 320
  ShowCursor = true
  [input0]
  Fire = Joy0-B0
  Jump = Joy0-B1
  Down = Joy0-H4
  Left = Joy0-H8
  Run = Joy0-B5
  Camlead = Joy0-B4
  Right = Joy0-H2
  Up = Joy0-H1
  Status = Joy0-B2
  Pogo = Joy0-B3
  
  '''
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_filter': 1}]
  tuple(
    list([
      'CGeniusExe',
      'dir="."',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_filter': 1}][config]
  '''
  [FileHandling]
  EnableLogfile = false
  SearchPath1 = /userdata/roms/cgenius
  SearchPath2 = /userdata/roms/cgenius/games
  [Video]
  aspect = 4:3
  fullscreen = false
  integerScaling = false
  filter = 1
  OGLfilter = nearest
  gameHeight = 200
  gameWidth = 320
  ShowCursor = false
  [input0]
  Fire = Joy0-B0
  Jump = Joy0-B1
  Down = Joy0-H4
  Left = Joy0-H8
  Run = Joy0-B5
  Camlead = Joy0-B4
  Right = Joy0-H2
  Up = Joy0-H1
  Status = Joy0-B2
  Pogo = Joy0-B3
  
  '''
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_quality': 'best'}]
  tuple(
    list([
      'CGeniusExe',
      'dir="."',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_quality': 'best'}][config]
  '''
  [FileHandling]
  EnableLogfile = false
  SearchPath1 = /userdata/roms/cgenius
  SearchPath2 = /userdata/roms/cgenius/games
  [Video]
  aspect = 4:3
  fullscreen = false
  integerScaling = false
  filter = none
  OGLfilter = best
  gameHeight = 200
  gameWidth = 320
  ShowCursor = false
  [input0]
  Fire = Joy0-B0
  Jump = Joy0-B1
  Down = Joy0-H4
  Left = Joy0-H8
  Run = Joy0-B5
  Camlead = Joy0-B4
  Right = Joy0-H2
  Up = Joy0-H1
  Status = Joy0-B2
  Pogo = Joy0-B3
  
  '''
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_render': '200'}]
  tuple(
    list([
      'CGeniusExe',
      'dir="."',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_render': '200'}][config]
  '''
  [FileHandling]
  EnableLogfile = false
  SearchPath1 = /userdata/roms/cgenius
  SearchPath2 = /userdata/roms/cgenius/games
  [Video]
  aspect = 4:3
  fullscreen = false
  integerScaling = false
  filter = none
  OGLfilter = nearest
  gameHeight = 200
  gameWidth = 320
  ShowCursor = false
  [input0]
  Fire = Joy0-B0
  Jump = Joy0-B1
  Down = Joy0-H4
  Left = Joy0-H8
  Run = Joy0-B5
  Camlead = Joy0-B4
  Right = Joy0-H2
  Up = Joy0-H1
  Status = Joy0-B2
  Pogo = Joy0-B3
  
  '''
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_render': '240'}]
  tuple(
    list([
      'CGeniusExe',
      'dir="."',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_render': '240'}][config]
  '''
  [FileHandling]
  EnableLogfile = false
  SearchPath1 = /userdata/roms/cgenius
  SearchPath2 = /userdata/roms/cgenius/games
  [Video]
  aspect = 4:3
  fullscreen = false
  integerScaling = false
  filter = none
  OGLfilter = nearest
  gameHeight = 240
  gameWidth = 320
  ShowCursor = false
  [input0]
  Fire = Joy0-B0
  Jump = Joy0-B1
  Down = Joy0-H4
  Left = Joy0-H8
  Run = Joy0-B5
  Camlead = Joy0-B4
  Right = Joy0-H2
  Up = Joy0-H1
  Status = Joy0-B2
  Pogo = Joy0-B3
  
  '''
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_render': '360'}]
  tuple(
    list([
      'CGeniusExe',
      'dir="."',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_render': '360'}][config]
  '''
  [FileHandling]
  EnableLogfile = false
  SearchPath1 = /userdata/roms/cgenius
  SearchPath2 = /userdata/roms/cgenius/games
  [Video]
  aspect = 4:3
  fullscreen = false
  integerScaling = false
  filter = none
  OGLfilter = nearest
  gameHeight = 360
  gameWidth = 640
  ShowCursor = false
  [input0]
  Fire = Joy0-B0
  Jump = Joy0-B1
  Down = Joy0-H4
  Left = Joy0-H8
  Run = Joy0-B5
  Camlead = Joy0-B4
  Right = Joy0-H2
  Up = Joy0-H1
  Status = Joy0-B2
  Pogo = Joy0-B3
  
  '''
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_render': '480'}]
  tuple(
    list([
      'CGeniusExe',
      'dir="."',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestCGeniusGenerator.test_generate_config[{'cgenius_render': '480'}][config]
  '''
  [FileHandling]
  EnableLogfile = false
  SearchPath1 = /userdata/roms/cgenius
  SearchPath2 = /userdata/roms/cgenius/games
  [Video]
  aspect = 4:3
  fullscreen = false
  integerScaling = false
  filter = none
  OGLfilter = nearest
  gameHeight = 480
  gameWidth = 640
  ShowCursor = false
  [input0]
  Fire = Joy0-B0
  Jump = Joy0-B1
  Down = Joy0-H4
  Left = Joy0-H8
  Run = Joy0-B5
  Camlead = Joy0-B4
  Right = Joy0-H2
  Up = Joy0-H1
  Status = Joy0-B2
  Pogo = Joy0-B3
  
  '''
# ---
# name: TestCGeniusGenerator.test_generate_controllers
  tuple(
    list([
      'CGeniusExe',
      'dir="."',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '''
        030000005e0400000a0b000005040000,real name 1,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,
        030000005e0400000a0b000005040000,real name 2,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,
        030000005e0400000a0b000005040000,real name 3,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,
        030000005e0400000a0b000005040000,real name 4,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,
        030000005e0400000a0b000005040000,real name 5,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,
      ''',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestCGeniusGenerator.test_generate_controllers[config]
  '''
  [FileHandling]
  EnableLogfile = false
  SearchPath1 = /userdata/roms/cgenius
  SearchPath2 = /userdata/roms/cgenius/games
  [Video]
  aspect = 4:3
  fullscreen = false
  integerScaling = false
  filter = none
  OGLfilter = nearest
  gameHeight = 200
  gameWidth = 320
  ShowCursor = false
  [input0]
  Fire = Joy0-B0
  Jump = Joy0-B1
  Down = Joy0-H4
  Left = Joy0-H8
  Run = Joy0-B5
  Camlead = Joy0-B4
  Right = Joy0-H2
  Up = Joy0-H1
  Status = Joy0-B2
  Pogo = Joy0-B3
  [input1]
  Fire = Joy1-B0
  Jump = Joy1-B1
  Down = Joy1-H4
  Left = Joy1-H8
  Run = Joy1-B5
  Camlead = Joy1-B4
  Right = Joy1-H2
  Up = Joy1-H1
  Status = Joy1-B2
  Pogo = Joy1-B3
  [input2]
  Fire = Joy2-B0
  Jump = Joy2-B1
  Down = Joy2-H4
  Left = Joy2-H8
  Run = Joy2-B5
  Camlead = Joy2-B4
  Right = Joy2-H2
  Up = Joy2-H1
  Status = Joy2-B2
  Pogo = Joy2-B3
  [input3]
  Fire = Joy3-B0
  Jump = Joy3-B1
  Down = Joy3-H4
  Left = Joy3-H8
  Run = Joy3-B5
  Camlead = Joy3-B4
  Right = Joy3-H2
  Up = Joy3-H1
  Status = Joy3-B2
  Pogo = Joy3-B3
  
  '''
# ---
# name: TestCGeniusGenerator.test_generate_existing
  tuple(
    list([
      'CGeniusExe',
      'dir="."',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestCGeniusGenerator.test_generate_existing[config]
  '''
  [FileHandling]
  EnableLogfile = false
  SearchPath1 = /userdata/roms/cgenius
  SearchPath2 = /userdata/roms/cgenius/games
  [Video]
  aspect = 4:3
  fullscreen = false
  integerScaling = false
  filter = none
  OGLfilter = nearest
  gameHeight = 200
  gameWidth = 320
  ShowCursor = false
  [input0]
  Fire = Joy0-B0
  Jump = Joy0-B1
  Down = Joy0-H4
  Left = Joy0-H8
  Run = Joy0-B5
  Camlead = Joy0-B4
  Right = Joy0-H2
  Up = Joy0-H1
  Status = Joy0-B2
  Pogo = Joy0-B3
  
  '''
# ---
# name: TestCGeniusGenerator.test_get_hotkeys_context
  dict({
    'keys': dict({
      'exit': list([
        'KEY_LEFTALT',
        'KEY_F4',
      ]),
      'menu': 'KEY_ESC',
      'pause': 'KEY_ESC',
      'restore_state': 'KEY_F9',
      'save_state': 'KEY_F6',
    }),
    'name': 'cgenius',
  })
# ---
